### PR TITLE
blockchain: Introduce `BlockchainMap` and `BlockchainKind`

### PR DESCRIPTION
**Pull request by leoyvens**
This introduces the `BlockchainMap` in which `Blockchain` instances are stored, keyed by their `BlockchainKind` and network. This is used in the entry points that need to call functions which are generic on `C: Blockchain`, which are: `InstanceManager::start_subgraph`, `SubgraphRegistrar::create_subgraph` and `IndexNodeResolver::resolve_subgraph_features`.

Part of #2732.

### CODE REVIEW

File: core/src/subgraph/registrar.rs [hunk 1]
```diff
@@ -271,46 +271,47 @@ where
...
+                    )
+                })?;
 
-        let chain = self
-            .chains
-            .get(&network_name)
-            .ok_or(SubgraphRegistrarError::NetworkNotSupported(
-                network_name.clone(),
-            ))?
-            .cheap_clone();
```
Comment (leoyvens) [inline core/src/subgraph/registrar.rs:294]: This block of code was moved to the `create_subgraph_version` free function.

File: graph/src/data/subgraph/mod.rs [hunk 2]
```diff
@@ -795,26 +795,22 @@ impl<C: Blockchain> UnvalidatedSubgraphManifest<C> {
...
         store: Arc<S>,
-    ) -> Result<
-        (SubgraphManifest<C>, Vec<SubgraphManifestValidationWarning>),
-        Vec<SubgraphManifestValidationError>,
-    > {
-        let (schemas, import_errors) = self.0.schema.resolve_schema_references(store.clone());
-        let validation_warnings = import_errors
-            .into_iter()
-            .map(SubgraphManifestValidationWarning::SchemaValidationWarning)
-            .collect();
```
Comment (leoyvens) [inline graph/src/data/subgraph/mod.rs:817]: These `import_errors` are not used anyways, so they are now ignored.

File: graph/src/data/subgraph/mod.rs [hunk 3]
```diff
@@ -921,40 +917,12 @@ impl<C: Blockchain> UnvalidatedSubgraphManifest<C> {
 
 impl<C: Blockchain> SubgraphManifest<C> {
     /// Entry point for resolving a subgraph definition.
-    /// Right now the only supported links are of the form:
-    /// `/ipfs/QmUmg7BZC1YP1ca66rRtWKxpXp77WgVHrnv263JtDuvs2k`
-    pub async fn resolve(
```
Comment (leoyvens) [inline graph/src/data/subgraph/mod.rs:926]: Removed because `resolve_from_raw` is now used in all cases.

File: node/src/main.rs [hunk 4]
```diff
@@ -205,15 +205,21 @@ async fn main() {
         let chain_head_update_listener = store_builder.chain_head_update_listener();
         let network_store = store_builder.network_store(idents);
 
-        let chains = Arc::new(networks_as_chains(
+        // To support the ethereum block ingestor, ethereum networks are referenced both by the
+        // `blockchain_map` and `ethereum_chains`. Future chains should be referred to only in
+        // `blockchain_map`.
+        let mut blockchain_map = BlockchainMap::new();
+        let ethereum_chains = networks_as_chains(
+            &mut blockchain_map,
```
Comment (leoyvens) [inline node/src/main.rs:213]: This hack is because I preferred to not refactor the block ingestor to use `BlockchainMap` since it is unlikely to be used for new chains, so `ethereum_chains` continues to exists.

File: graph/src/data/subgraph/mod.rs [hunk 5]
```diff
@@ -1024,6 +992,7 @@ impl<C: Blockchain> UnresolvedSubgraphManifest<C> {
         self,
         resolver: &impl LinkResolver,
         logger: &Logger,
+        max_spec_version: semver::Version,
```
Comment (leoyvens) [inline graph/src/data/subgraph/mod.rs:995]: Adding this is actually unrelated to this PR, it's just improving the tests to no need the GRAPH_MAX_SPEC_VERSION env var set which was introduced in https://github.com/graphprotocol/graph-node/pull/2682.

File: graph/src/blockchain/mod.rs [hunk 6]
```diff
@@ -305,3 +310,73 @@ pub trait RuntimeAdapter<C: Blockchain>: Send + Sync {
...
+impl FromStr for BlockchainKind {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "ethereum" => Ok(BlockchainKind::Ethereum),
+            _ => Err(anyhow!("unknown blockchain kind {}", s)),
+        }
+    }
+}
```
Comment (tilacog) [inline graph/src/blockchain/mod.rs:339]: How do you feel about delegating this to `serde::plain`, just like we did for [subgraph features](https://github.com/graphprotocol/graph-node/blob/master/graph/src/data/subgraph/features.rs#L36-L51)?

While the current implementation works flawlessly, if we use `serde` here we wouldn't need to write any more code when introducing new variants.

File: server/index-node/src/resolver.rs [hunk 7]
```diff
@@ -379,7 +396,9 @@ where
             }
 
             // The top-level `indexingStatusForPendingVersion` field
-            (None, "subgraphFeatures") => self.resolve_subgraph_features(arguments),
+            (None, "subgraphFeatures") => {
+                graph::block_on(self.resolve_subgraph_features(arguments))
+            }
```
Comment (tilacog) [inline server/index-node/src/resolver.rs:401]: Way better! :clap:

**Review by tilacog — APPROVED** [review]

File: graph/src/blockchain/mod.rs (hunk 6 above)
Comment (leoyvens) [reply graph/src/blockchain/mod.rs:339]: I don't have much of a preference so I'll keep it as it is.
